Update the docs for the kernel ROM chiplet

grjte opened the pull request:
This PR adds a docs page for the kernel ROM chiplet and updates the chiplets docs more generally according to the recent changes made in #928 and #939

- [x] add a docs page for the kernel ROM chiplet
- [x] update the main chiplets page
- [x] write new virtual table section for main chiplets page
- [x] check the degrees of the chiplets to be sure they're correct
- [x] create a new image for the chiplets trace (main chiplets page)
- [x] create an image for the kernel ROM trace (kernel ROM chiplet page)
- [x] main design page: add Kernel ROM chiplet to the VM components section.
- [x] main design page: Update VM execution trace picture and associated description. Here, besides the chiplets segment, would be great to update the system and program decoder segments (by using the right number of columns).
- [x] update the hasher sibling table docs
- [x] update the block hash computation constraint for SYSCALL
- [x] update the multiset checks docs
- [x] fix issue with last row constraints description on main chiplets page

grjte commented:
@bobbinth I need to finalize the new images, but the other docs changes are all finished, so it would be great to get your feedback on them (and whether I missed anything)

bobbinth commented on docs/src/design/chiplets/main.md:
```diff
@@ -81,9 +58,11 @@ This requires the following adjustments for each chiplet.
 ## Operation labels
 
-Each operation supported by the chiplets is given a unique identifier to ensure that the requests and responses sent to $b_{chip}$ are indeed processed by the intended chiplet for that operation and that chiplets which support more than one operation execute the correct one.
+Each operation supported by the chiplets is given a unique identifier to ensure that the requests and responses sent to the [chiplets bus](#chiplets-bus) ($b_{chip}$) are indeed processed by the intended chiplet for that operation and that chiplets which support more than one operation execute the correct one.
```
There is a `\` missing on line 81 below in the "Combined Flag" column.

bobbinth commented on docs/src/design/chiplets/main.md:
```diff
@@ -154,6 +140,64 @@ The following constraints ensure that the chiplets are stacked correctly by rest
+- enforcing a boundary constraint against the last row for the value of all of the kernel ROM procedures
+- using the following transition constraint to enforce that when the active chiplet changes to the kernel ROM chiplet the value is $1$.
+
+> $$
+s_0 \cdot s_1 \cdot (s'_2 - s_2) \cdot (1 - vt'_{chip}) = 0 \text{ | degree} = 4
+$$
```
I'm not sure this is necessary - but let's keep it. We should probably make a note somewhere that this is something we can potentially optimize away.

grjte replied:
Hm. I thought about it, but to me it seems that we need to constrain the start & end values of each virtual table, and if we combine the virtual tables, then we would still need to constrain the start/end of those "chunks" of the combined table

bobbinth replied:
My thinking was that we already verify that the value in the column is `ONE` at the end of the hash chiplet segment, and no other chiplet can change the value - so, it should remain `ONE` at the start of kernel ROM chiplet. But again, I think it is fine to keep for now and we can revisit later if needed.

bobbinth reviewed (changes requested):
Looks great! Thank you! I left a few small comments inline.

We should also probably update the main design page to do the following:
- Add Kernel ROM chiplet to the VM components section.
- Update VM execution trace picture and associated description. Here, besides the chiplets segment, would be great to update the system and program decoder segments (by using the right number of columns).

grjte commented on docs/src/design/decoder/constraints.md:
```diff
@@ -168,13 +184,13 @@ $$
 > $$
-p_0' \cdot (u_{ctrli} + u_{span} + u_{respan} + u_{end} + \\
-1 - (f_{ctrli} + f_{span} + f_{respan} + f_{end})) = p_0
+b_{chip}' \cdot (u_{ctrli} + u_{syscall} + u_{span} + u_{respan} + u_{end} + \\
+1 - (f_{ctrli} + f_{span} + f_{respan} + f_{end})) = b_{chip}
```
Yes! Good catch, thank you

grjte commented on docs/src/design/decoder/constraints.md:
```diff
@@ -128,10 +130,10 @@ $$
 $$
 
-In the above, $f_{ctrli}$ is set to $1$ when a control flow operation that signifies the initialization of a control block is being executed on the VM.  Otherwise, it is set to $0$.
+In the above, $f_{ctrli}$ is set to $1$ when a control flow operation that signifies the initialization of a control block is being executed on the VM.  Otherwise, it is set to $0$. An exception is made for the `SYSCALL` operation. Although it also signifies the initialization of a control block, it must additionally send a procedure access request to the [kernel ROM chiplet](../chiplets/kernel_r ...
```
Debatable, I think. I left it there because it's true that it initializes a control block and everything described in that paragraph needs to be done for it. If I remove it from that paragraph, then we need to describe it again for `SYSCALL`, which we could do. I tried to address it with the paragraph below instead, but if you want me to remove it from 146 I can

bobbinth replied:
No - let's keep it then.

grjte commented:
Thanks for your review @bobbinth! I've addressed your comments and finished adding the new images. I also addressed the two main design points you noted:

> We should also probably update the main design page to do the following:
> 
>     * Add Kernel ROM chiplet to the VM components section.
> 
>     * Update VM execution trace picture and associated description. Here, besides the chiplets segment, would be great to update the system and program decoder segments (by using the right number of columns).

I adjusted the description of system to be more accurate, but didn't add docs for the way it currently works, since we're planning to change that. I believe the docs for the decoder are also a bit outdated, and I didn't address these either, since we're planning to make some changes here as well
